syntax = "proto3";

package server;

option go_package = "wyron-panel/api/grpc/proto";

import "google/protobuf/empty.proto";

message Server {
  string id = 1;
  string display_name = 2;
  string address = 3;
  string username = 4;
  int64 created_at = 5;
  repeated WireGuardInterface interfaces = 6;
}

message WireGuardInterface {
  string name = 1;
  string display_name = 2;
  string subnet = 3;
  string dns = 4;
  string endpoint = 5;
  int32 port = 6;
  string public_key = 7;
  int64 created_at = 8;
}

message UpdateServerRequest {
  string display_name = 1;
  string address = 2;
  string username = 3;
  string password = 4;
}

message ListServersResponse {
  repeated Server servers = 1;
}

message ServerIDRequest {
  string id = 1;
}

message UpdateInterfaceResponse {
  WireGuardInterface interface = 1;
}

message InterfaceRequest {
  string server_id = 1;
  string name = 2;
  string display_name = 3;
  string endpoint = 4;
  string dns = 5;
}

service ServerService {
  rpc List (google.protobuf.Empty) returns (ListServersResponse);
  rpc Get (ServerIDRequest) returns (Server);
  rpc Update (UpdateServerRequest) returns (Server);
  rpc Delete (ServerIDRequest) returns (google.protobuf.Empty);
  rpc UpdateInterface (InterfaceRequest) returns (UpdateInterfaceResponse);
  rpc DeleteInterface (InterfaceRequest) returns (google.protobuf.Empty);
}
